keydown_util: Add is_enter_event check for IME edge cases.

aero31aero opened the pull request:
We now explicitly check whether the keycode was 13 and not
some IME selection enter event that also use e.key = "Enter"
behavior.

Fixes #22062.

<!-- Describe your pull request here.-->

Fixes: <!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented:
Have you successfully tested this? Also looks like it needs a minor rebase.

timabbott commented on static/js/keydown_util.ts:
```diff
@@ -31,3 +31,11 @@ export function handle(opts: {
+    // we don't trigger our Enter behaviour on those
+    // composing events.
+    return event.which === 13;
+}
```
Do we want to check `event.key` also? I don't have a great sense for how this is supposed to work.

Another note -- this PR might read better with a first commit that just extracts `is_enter_event` (with the implementation being still `e.key`) and then a commit that changes the behavior.

chrisbobbe replied:
> Another note -- this PR might read better with a first commit that just extracts `is_enter_event` (with the implementation being still `e.key`) and then a commit that changes the behavior.

Agreed. 👍

I'd be interested in trying the [original proposal](https://chat.zulip.org/#narrow/stream/9-issues/topic/enter.20to.20send.20CJK.20IME.20error.20.2322062/near/1382010):

> Can we make use of `.isComposing` on `KeyboardEvent`s? https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/isComposing

> That should tell us if a `KeyboardEvent` was emitted as part of an [IME composing session](https://developer.mozilla.org/en-US/docs/Glossary/Input_method_editor). When that's the case, we shouldn't let Enter send the Zulip message, right?

@aero31aero, I understand from [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/enter.20to.20send.20CJK.20IME.20error.20.2322062/near/1427471) that you don't have a Mac available for testing, so I'll follow through with my offer in May:

> I'm hoping someone who's more familiar with the web front-end code than I am can look at zulip/zulip@bf179b7d2. If that person doesn't have a Mac or otherwise has difficulty reproducing the issue, I can help with manual testing.

I plan to test both the current implementation and one that uses `isComposing`.

chrisbobbe commented:
Thanks, @aero31aero!

This revision seems to fix #22062. I haven't managed to reproduce the issue on this branch.

Some thoughts for improvement (Tim, you might have different thoughts):

- I think `is_enter_event` is the wrong name: we're asking it to return _false_ for some "enter events," i.e., `KeyboardEvent`s from the user pressing Enter, with `.key` and `.code` "Enter". In particular, "enter events" that happen during an IME session.
  - Perhaps `is_non_IME_enter_event`?
  - Or maybe the more general `is_non_IME_event`, with the implementation changed to match. That would be useful [if we find non-"Enter" events where the IME status is relevant](https://chat.zulip.org/#narrow/stream/9-issues/topic/enter.20to.20send.20CJK.20IME.20error.20.2322062/near/1383627).
- Can we give a more specific type than `any` for the function's param? If we can say it's a `KeyboardEvent`, or (I'm guessing here) a `JQuery.KeyDownEvent` or [one of its friends](https://api.jquery.com/category/events/keyboard-events/), it'll help both the function's interface and its implementation:
  - Its interface, because callers can be more confident that they're passing an input that the function will actually handle.
  - Its implementation, because we can answer questions like, "Is the `.which` check using a [deprecated field that might be dropped](https://developer.mozilla.org/en-US/docs/Web/API/UIEvent/which), or is it using [jQuery's recommended, 'normalized' field](https://api.jquery.com/keydown/)?" just by looking at the function's definition.
- If we want to use [`.isComposing`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/isComposing), I think we can only get it from a `KeyboardEvent`. That might be what gets passed to this new helper function; or, if that's a jQuery wrapper for the event, maybe we can get one from that wrapper at [`.originalEvent`](https://api.jquery.com/category/events/event-object/). In either case, we should type the function's param more specifically, as I suggested in the previous bullet, so we know we're looking for `.isComposing` in the right place.
- In the code comment, let's write down more things we know that might be relevant to readers:
  - "229 is a special value set for a `keyCode` relating to an event that has been processed by an IME." That's [a quote from MDN](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event), and we can include that link.
  - [`KeyboardEvent.isComposing`](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/isComposing) seems like the preferred, explicit way to tell whether you're in an IME session. We can include that MDN link too.

-----

In case it's helpful, I'll quote MDN's example for how to solve our problem. One of the links is broken; this is from [here](https://developer.mozilla.org/en-US/docs/Web/API/Element/keydown_event):

> Since Firefox 65, the `keydown` and [`keyup`](/en-US/docs/Web/API/Element/keyup_event) events are now fired during IME composition, to improve cross-browser compatibility for CJKT users ({{bug(354358)}}. To ignore all `keydown` events that are part of composition, do something like this (229 is a special value set for a `keyCode` relating to an event that has been processed by an IME):
> 
> ```js
> eventTarget.addEventListener("keydown", (event) => {
>   if (event.isComposing || event.keyCode === 229) {
>     return;
>   }
>   // do something
> });
> ```

timabbott commented:
For naming, I think we want a function with a name like `should_process_as_enter(e)` for what we call from places. The callers shouldn't need to know anything about IME and I think it'd obscure the code to have these 20 call points all mention it.

timabbott commented:
I'm pretty sure it's a `KeyboardEvent`.

timabbott commented:
@aero31aero do you have time to finish this? If not, I'd like us to have someone else pick it up (perhaps @N-Shar-ma, since she's been working in the area)

aero31aero commented:
Ping. Apparently marking the PR as ready-for-review doesn't trigger any notifications.

alya commented:
@timabbott This PR has been updated, so marking it as "integration review" so that we don't lose track. That's all I know about this one. ;)

timabbott commented:
@chrisbobbe would it be convenient for you to repeat testing on this version?

chrisbobbe commented:
In the current revision—

2ba8ca53c42b5f3b9ce0a9347865fc9306f847fb keydown_util: Extract all 'Enter' events to one place.
dd1a5e44669dfeb572f192c980ffbebdef855840 keydown_util: Check Enter events properly for IME edge cases.

—I actually don't see any sign that @aero31aero considered my messages on this PR page, or Tim's message at https://github.com/zulip/zulip/pull/22811#issuecomment-1232259165. @aero31aero, were there updated commits you forgot to send, or GitHub comments you forgot to submit? 🙂 I'll be happy to spend my time testing a revision we're more confident about, or someone else can test this one with the reproduction recipe I've linked to.

chrisbobbe commented:
Tim says:

> I'm pretty sure it's a `KeyboardEvent`.

This sounds like a suggestion to try the `KeyboardEvent` type in place of `any` for the new helper function's param. I'd like to know if all callers satisfy that or not, as I mentioned. But in any case, [TypeScript's `any` defeats type-checking](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#any), so we should avoid it when possible.

alya commented:
@aero31aero I'm removing the integration review label until you get a chance to address @chrisbobbe 's feedback. Thanks!

aero31aero commented:
@chrisbobbe Sorry I didn't address your specific concerns with this PR earlier, I missed that comment.

Regarding the name suggestions, I feel `is_enter_event` concisely explains what the function does. `is_non_IME_event` and variations, while technically more precise, feel clunky to read and type.

Also, I'd found some similar questions on SO as well as my testing led me to believe `event.which === 13` is foolproof, thus doesn't require the `isComposing` check. While that remains true, I've read up a bit more and since `event.which` is now deprecated, it makes sense to use something like:  `if(!event.isComposing && event.key === "Enter")` instead.

I've implemented this behavior instead, which also resulted in cleaned-up tests. If you could, I'd appreciate you doing a manual test of this PR now.

timabbott commented:
Yeah I agree we want a name like `is_enter_event` that helps a reader of the 20 places that will call this function understand it correctly as "Should this key press be processed as an Enter event".

timabbott commented:
I'm going to test-deploy this on chat.zulip.org, so we can get some burn-in on with whether it causes any apparent regressions there.

legopin commented:
Hi, I tested this on chat.zulip.org.

The issue is fixed in the compose box's topic field.
Issue still at topic edit in message stre
<img width="510" alt="Screen Shot 2022-10-29 at 12 18 33 PM" src="https://user-images.githubusercontent.com/15863509/198813697-eaffe604-5b9e-4109-9a03-617fbbb552d3.png">
am

aero31aero commented:
Thanks @legopin, good spot! I'd missed two 'Enter' comparisons; the last two commits should fix them.

chrisbobbe reviewed (dismissed):
Thanks @aero31aero, and thanks @legopin for your testing!

I've just tested the edit-topic input—
<img width="362" alt="image" src="https://user-images.githubusercontent.com/22248748/199111826-a29f87a1-e4d3-48de-8d91-673a358cfbd6.png">
—and I didn't see an issue where "Enter" causes the topic edit to get prematurely submitted during an IME session.

However, I did see a different issue. The topic autocomplete dropdown, partly visible in that screenshot, responds to an IME enter event by performing an autocomplete with whatever item is currently selected, instead of ignoring the event as it should. So I suspect there's another place where we should be calling the new helper function.

The last two commits aren't best described as "refactors" (and note the spelling). They intentionally change how the app actually behaves, by fixing prominent bugs. The commit in this PR that just pulls out a helper function is a good example of a refactor.

For the new function's name, I personally prefer Tim's suggestion `should_process_as_enter` over `is_enter_event` because `is_enter_event` tells a lie, but I'll defer to Tim on this. 🙂

chrisbobbe commented:
Oh also, Tim said at https://github.com/zulip/zulip/pull/22811#issuecomment-1234745768:

> I'm pretty sure it's a `KeyboardEvent`.

Tim, if there are any cases where you think `is_enter_event` is actually receiving a `KeyboardEvent`, then we should investigate, because the fix won't work in those cases. This revision has `is_enter_event` accept a `JQuery.KeyDownEvent`, and apparently that doesn't raise any complaints from TypeScript.

aero31aero commented:
@chrisbobbe please test with the latest commit as well. I couldn't do any testing beyond "the app loads fine and there are no console errors".

As for the changes, I:

- Moved the commits around so now they are indeed 'refactors'.
- Moved the comment to the right commit.

chrisbobbe commented:
Thanks for the revision! The issue with the topic-autocomplete dropdown seems to be fixed, and I didn't discover any additional bugs in my testing just now. 🙂

timabbott commented:
I merged all but the last commit, after edits mainly on the comments, and pushed the last commit back here.

I'm not certain whether or not the `-13` approach is a good one; I'd love to see some notes on the reasoning behind that design choice, but mainly I've just not had the time to read it fully.

timabbott commented:
I rewrote the remaining commit to invert the situation in which we create a `-13` event so that those are the IME events, not the normal `Enter` events; I think this implementation feels cleaner, since in particular we don't mutate the original event object (which could have consequences outside the typeahead library).

chrisbobbe commented:
Hmm. With focus in the topic input in the compose box, keypresses are causing red error boxes to appear:

<img width="1435" alt="image" src="https://user-images.githubusercontent.com/22248748/202050613-a104e323-5b15-473f-a4d0-772f5f6eddac.png">

I pressed `1` and three of them appeared; I pressed "delete" and two of them appeared; I pressed the keychord to screenshot part of the screen, and a bunch more appeared.

At the commit just before [b53fbda](https://github.com/zulip/zulip/commit/b53fbda1c96e1ef841fff414da71f740dc100daf), that problem isn't happening.

timabbott commented:
Yep, sorry about that; fixed, can you test again?

chrisbobbe commented:
Hmm, has something else in this area changed since https://github.com/zulip/zulip/pull/22811#issuecomment-1301158770 ? In following the steps to reproduce the [topic-autocomplete issue](https://github.com/zulip/zulip/pull/22811#pullrequestreview-1162512379), I was surprisingly unable to reproduce a situation like in my [screenshot](https://github.com/zulip/zulip/pull/22811#pullrequestreview-1162512379), where both the topic autocomplete and the IME menu were shown at the same time. So I can't yet confirm that in such a situation, if it still exists, only the IME menu and not the topic autocomplete would respond to enter events.

Here's what I tried:

1. There are some messages with the topic "test 🐎".
2. There are some different messages with the topic "test 123".
3. In the topic header for the "test 123" messages, I clicked the pencil icon to edit the topic, and deleted the " 123", so that "test 🐎" showed up in the topic autocomplete.
4. In Zhuyin input mode, I typed "a", to start the sequence "a83" that would trigger the menu that offers 🐎, 馬, etc.
5. But at that point, surprisingly, the topic autocomplete disappeared.

It may be that a Google Chrome update is a factor… Ooh, I did upgrade to macOS Ventura 13.0 recently, and that could be a factor because the IME interface is presented by macOS.

@aero31aero's revision of this commit did pass my testing (https://github.com/zulip/zulip/pull/22811#issuecomment-1301158770), so if you're confident in the code, I guess go ahead and merge? 🤷

timabbott commented:
Interesting; I could easily see an OS improvement hiding these `isComposing` enter events, since it does basically seem like an OS/browser interaction bug that we have to think about them at all. 

This failed puppeteer tests, though the exception looks like the thing we had before the last update, so hopefully it'll pass after a rebase.

chrisbobbe commented:
> Interesting; I could easily see an OS improvement hiding these isComposing enter events, since it does basically seem like an OS/browser interaction bug that we have to think about them at all.

Possibly, although I should underline that my testing today didn't suggest anything about the `isComposing` enter events being hidden or not.

This bit—

> @aero31aero's revision of this commit did pass my testing (https://github.com/zulip/zulip/pull/22811#issuecomment-1301158770), so if you're confident in the changes you've made to that, I guess go ahead and merge? 🤷

—referred to the testing I did at that time, 13 days ago, and that was before my macOS upgrade (and there may have been a Chrome auto-upgrade somewhere in there too).

timabbott commented:
OK; I'll merge and deploy to chat.zulip.org, and ask for an additional test from the other folks who could reproduce it.

aero31aero commented:
I like this way of rewriting the patch a lot. Although it looks like its time to upgrade `third/typeahead` to `js/typeahead` and make the code look and feel less hacky.
